### PR TITLE
Send pod name/ns to nodeagent

### PR DESCRIPTION
**Pull request by jayanthvn**
**What type of PR is this?**

Feature

**Which issue does this PR fix**:
N/A

**What does this PR do / Why do we need it**:
In certain scenarios we would need to send pod name/ns to node agent

**If an issue # is not available please add repro steps and logs from IPAMD/CNI showing the issue**:
N/A

**Testing done on this change**:

Yes

Failed call to NA and kubelet will send delete so need to call cleanup... -

```
{"level":"info","ts":"2024-02-14T17:28:56.972Z","caller":"routed-eni-cni-plugin/cni.go:196","msg":"Received add network response from ipamd for container 0f281751103d315a94be105eb2c37e7e4f030cbd51dca3d72803306722fbafe5 interface eth0: Success:true IPv4Addr:\"192.168.54.253\" DeviceNumber:2 VPCv4CIDRs:\"192.168.0.0/16\" NetworkPolicyMode:\"strict\" "}
{"level":"debug","ts":"2024-02-14T17:28:56.972Z","caller":"routed-eni-cni-plugin/cni.go:296","msg":"SetupPodNetwork: hostVethName=eni4a5c3305307, contVethName=eth0, netnsPath=/var/run/netns/cni-2a7be20f-ffff-d303-9ea0-783cdf93c4c9, v4Addr=192.168.54.253/32, v6Addr=<nil>, deviceNumber=2, mtu=9001"}
{"level":"debug","ts":"2024-02-14T17:28:57.008Z","caller":"driver/driver.go:237","msg":"Successfully set IPv6 sysctls on hostVeth eni4a5c3305307"}
{"level":"debug","ts":"2024-02-14T17:28:57.008Z","caller":"driver/driver.go:253","msg":"Successfully setup container route, containerAddr=192.168.54.253/32, hostVeth=eni4a5c3305307, rtTable=main"}
{"level":"debug","ts":"2024-02-14T17:28:57.008Z","caller":"driver/driver.go:253","msg":"Successfully setup toContainer rule, containerAddr=192.168.54.253/32, rtTable=main"}
{"level":"debug","ts":"2024-02-14T17:28:57.008Z","caller":"driver/driver.go:253","msg":"Successfully setup fromContainer rule, containerAddr=192.168.54.253/32, rtTable=3"}
{"level":"debug","ts":"2024-02-14T17:28:57.008Z","caller":"routed-eni-cni-plugin/cni.go:196","msg":"Using dummy interface: {Name:dummy4a5c3305307 Mac:0 Sandbox:2}"}
{"level":"error","ts":"2024-02-14T17:28:57.013Z","caller":"routed-eni-cni-plugin/cni.go:196","msg":"Failed to setup default network policy Pod Name websocket-client-7f65cc6544-sbwwz and NameSpace default: GRPC returned - rpc error: code = Unavailable desc = error reading from server: EOF Network policy agent returned - <nil>"}
{"level":"info","ts":"2024-02-14T17:28:57.030Z","caller":"routed-eni-cni-plugin/cni.go:119","msg":"Constructed new logger instance"}
{"level":"debug","ts":"2024-02-14T17:28:57.031Z","caller":"routed-eni-cni-plugin/cni.go:381","msg":"Prev Result: <nil>\n"}
{"level":"info","ts":"2024-02-14T17:28:57.031Z","caller":"routed-eni-cni-plugin/cni.go:381","msg":"Received CNI del request: ContainerID(0f281751103d315a94be105eb2c37e7e4f030cbd51dca3d72803306722fbafe5) Netns(/var/run/netns/cni-2a7be20f-ffff-d303-9ea0-783cdf93c4c9) IfName(eth0) Args(IgnoreUnknown=1;K8S_POD_NAMESPACE=default;K8S_POD_NAME=websocket-client-7f65cc6544-sbwwz;K8S_POD_INFRA_CONTAINER_ID=0f281751103d315a94be105eb2c37e7e4f030cbd51dca3d72803306722fbafe5;K8S_POD_UID=a2385f19-a5e7-44fa-b2ab-9009213c1ea2) Path(/opt/cni/bin) argsStdinData({\"cniVersion\":\"0.4.0\",\"mtu\":\"9001\",\"name\":\"aws-cni\",\"pluginLogFile\":\"/var/log/aws-routed-eni/plugin.log\",\"pluginLogLevel\":\"DEBUG\",\"podSGEnforcingMode\":\"strict\",\"type\":\"aws-cni\",\"vethPrefix\":\"eni\"})"}
{"level":"info","ts":"2024-02-14T17:28:57.033Z","caller":"routed-eni-cni-plugin/cni.go:381","msg":"Received del network response from ipamd for pod websocket-client-7f65cc6544-sbwwz namespace default sandbox 0f281751103d315a94be105eb2c37e7e4f030cbd51dca3d72803306722fbafe5: Success:true IPv4Addr:\"192.168.54.253\" DeviceNumber:2 "}
{"level":"debug","ts":"2024-02-14T17:28:57.033Z","caller":"routed-eni-cni-plugin/cni.go:503","msg":"TeardownPodNetwork: containerAddr=192.168.54.253/32, deviceNumber=2"}
{"level":"debug","ts":"2024-02-14T17:28:57.033Z","caller":"driver/driver.go:267","msg":"Successfully deleted toContainer rule, containerAddr=192.168.54.253/32, rtTable=main"}
{"level":"debug","ts":"2024-02-14T17:28:57.033Z","caller":"driver/driver.go:267","msg":"Successfully deleted fromContainer rule, containerAddr=192.168.54.253/32, rtTable=3"}
{"level":"debug","ts":"2024-02-14T17:28:57.033Z","caller":"driver/driver.go:267","msg":"Successfully deleted container route, containerAddr=192.168.54.253/32, rtTable=main"}
```

Successful call to NA -

```
{"level":"info","ts":"2024-02-14T08:23:06.969Z","caller":"routed-eni-cni-plugin/cni.go:196","msg":"Received CNI add request: ContainerID(fc1a62966a5312f7e5349733fbf741db1eb3790f7cfb336223c985d0e8b510ca) Netns(/var/run/netns/cni-7481806d-2e06-7ac7-1a02-7c3485e43a03) IfName(eth0) Args(IgnoreUnknown=1;K8S_POD_NAMESPACE=default;K8S_POD_NAME=websocket-client-7f65cc6544-jfjdh;K8S_POD_INFRA_CONTAINER_ID=fc1a62966a5312f7e5349733fbf741db1eb3790f7cfb336223c985d0e8b510ca;K8S_POD_UID=929314c7-93af-4267-b629-766c05af7ae4) Path(/opt/cni/bin) argsStdinData({\"cniVersion\":\"0.4.0\",\"mtu\":\"9001\",\"name\":\"aws-cni\",\"pluginLogFile\":\"/var/log/aws-routed-eni/plugin.log\",\"pluginLogLevel\":\"DEBUG\",\"podSGEnforcingMode\":\"strict\",\"type\":\"aws-cni\",\"vethPrefix\":\"eni\"})"}
{"level":"debug","ts":"2024-02-14T08:23:06.969Z","caller":"routed-eni-cni-plugin/cni.go:196","msg":"Prev Result: <nil>\n"}
{"level":"debug","ts":"2024-02-14T08:23:06.969Z","caller":"routed-eni-cni-plugin/cni.go:196","msg":"MTU value set is 9001:"}
{"level":"info","ts":"2024-02-14T08:23:06.973Z","caller":"routed-eni-cni-plugin/cni.go:196","msg":"Received add network response from ipamd for container fc1a62966a5312f7e5349733fbf741db1eb3790f7cfb336223c985d0e8b510ca interface eth0: Success:true IPv4Addr:\"192.168.58.56\" DeviceNumber:1 VPCv4CIDRs:\"192.168.0.0/16\" NetworkPolicyMode:\"strict\" "}
{"level":"debug","ts":"2024-02-14T08:23:06.973Z","caller":"routed-eni-cni-plugin/cni.go:296","msg":"SetupPodNetwork: hostVethName=eniabdd087b487, contVethName=eth0, netnsPath=/var/run/netns/cni-7481806d-2e06-7ac7-1a02-7c3485e43a03, v4Addr=192.168.58.56/32, v6Addr=<nil>, deviceNumber=1, mtu=9001"}
{"level":"debug","ts":"2024-02-14T08:23:07.020Z","caller":"driver/driver.go:237","msg":"Successfully set IPv6 sysctls on hostVeth eniabdd087b487"}
{"level":"debug","ts":"2024-02-14T08:23:07.020Z","caller":"driver/driver.go:253","msg":"Successfully setup container route, containerAddr=192.168.58.56/32, hostVeth=eniabdd087b487, rtTable=main"}
{"level":"debug","ts":"2024-02-14T08:23:07.020Z","caller":"driver/driver.go:253","msg":"Successfully setup toContainer rule, containerAddr=192.168.58.56/32, rtTable=main"}
{"level":"debug","ts":"2024-02-14T08:23:07.020Z","caller":"driver/driver.go:253","msg":"Successfully setup fromContainer rule, containerAddr=192.168.58.56/32, rtTable=2"}
{"level":"debug","ts":"2024-02-14T08:23:07.020Z","caller":"routed-eni-cni-plugin/cni.go:196","msg":"Using dummy interface: {Name:dummyabdd087b487 Mac:0 Sandbox:1}"}
{"level":"debug","ts":"2024-02-14T08:23:07.128Z","caller":"routed-eni-cni-plugin/cni.go:196","msg":"Network Policy agent returned Success : true"}
```

Enabling strict mode even coredns will be blocked and egress from client pod will be blocked-

```
{"level":"info","ts":"2024-02-14T08:26:49.020Z","logger":"ebpf-client","msg":"Flow Info:  ","Src IP":"192.168.58.56","Src Port":34043,"Dest IP":"10.100.0.10","Dest Port":53,"Proto":"UDP","Verdict":"DENY"}
{"level":"info","ts":"2024-02-14T08:26:49.020Z","logger":"ebpf-client","msg":"Flow Info:  ","Src IP":"192.168.58.56","Src Port":48442,"Dest IP":"10.100.0.10","Dest Port":53,"Proto":"UDP","Verdict":"DENY"}
```

Upon enabling egress on client pod and ingress on coredns pods -

```
"Verdict":"ACCEPT"}
{"level":"info","ts":"2024-02-14T08:33:31.010Z","logger":"ebpf-client","msg":"Flow Info:  ","Src IP":"192.168.58.56","Src Port":58073,"Dest IP":"10.100.0.10","Dest Port":53,"Proto":"UDP","Verdict":"ACCEPT"}
{"level":"info","ts":"2024-02-14T08:33:31.010Z","logger":"ebpf-client","msg":"Flow Info:  ","Src IP":"192.168.58.56","Src Port":43721,"Dest IP":"10.100.0.10","Dest Port":53,"Proto":"UDP","Verdict":"ACCEPT"}
{"level":"info","ts":"2024-02-14T08:33:31.011Z","logger":"ebpf-client","msg":"Flow Info:  ","Src IP":"192.168.58.56","Src Port":33020,"Dest IP":"10.100.58.78","Dest Port":80,"Proto":"TCP","Verdict":"ACCEPT"}
```

But client to server is blocked, since server hasn't allowed ingress -

```
"level":"info","ts":"2024-02-14T08:34:02.540Z","logger":"ebpf-client","msg":"Flow Info:  ","Src IP":"192.168.58.56","Src Port":33020,"Dest IP":"192.168.50.232","Dest Port":8080,"Proto":"TCP","Verdict":"DENY"} 

default       websocket-client-7f65cc6544-jfjdh                       0/1     CrashLoopBackOff   5 (72s ago)    13m     192.168.58.56    ip-192-168-50-132.us-west-2.compute.internal   <none>           <none>
default       websocket-server-64ddc8864-zrl9v                        1/1     Running            0              6m46s   192.168.50.232   ip-192-168-50-132.us-west-2.compute.internal   <none>           <none>
```

Upon allowing client to server -

```
{"level":"info","ts":"2024-02-14T08:38:06.864Z","logger":"ebpf-client","msg":"Flow Info:  ","Src IP":"192.168.52.244","Src Port":37262,"Dest IP":"192.168.50.232","Dest Port":8080,"Proto":"TCP","Verdict":"ACCEPT"}
```

Even pod will be running -

```
default       websocket-client-7f65cc6544-xrt2z                       1/1     Running   0              20s     192.168.52.244   ip-192-168-50-132.us-west-2.compute.internal   <none>           <none>
default       websocket-server-64ddc8864-zrl9v                        1/1     Running   0              13m     192.168.50.232   ip-192-168-50-132.us-west-2.compute.internal   <none>           <none>
```
**Will this PR introduce any new dependencies?**:

No

**Will this break upgrades or downgrades? Has updating a running cluster been tested?**:
No

**Does this change require updates to the CNI daemonset config files to work?**:

No

**Does this PR introduce any user-facing change?**:

No

```release-note
New env variable - NETWORK_POLICY_ENFORCING_MODE:   [none, strict, standard]
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cmd/routed-eni-cni-plugin/cni.go [hunk 1]
```diff
@@ -275,6 +277,38 @@ func add(args *skel.CmdArgs, cniTypes typeswrapper.CNITYPES, grpcClient grpcwrap
 	// dummy interface is appended to PrevResult for use during cleanup
 	result.Interfaces = append(result.Interfaces, dummyInterface)
 
+	if r.NetworkPolicyMode == "strict" {
```
Comment (jdn5126) [inline cmd/routed-eni-cni-plugin/cni.go:280]: nit: let's either add a comment above this section or move it into a new function. We should explain what strict mode is and why CNI is waiting

File: cmd/routed-eni-cni-plugin/cni.go [hunk 2]
```diff
@@ -275,6 +277,38 @@ func add(args *skel.CmdArgs, cniTypes typeswrapper.CNITYPES, grpcClient grpcwrap
...
+		// Set up a connection to the ipamD server.
+		npConn, err := grpcClient.Dial(npAgentAddress, grpc.WithTransportCredentials(insecure.NewCredentials()))
+		if err != nil {
+			log.Errorf("Failed to connect to network policy agent: %v", err)
+			return errors.Wrap(err, "add cmd: failed to connect to network policy agent backend server")
```
Comment (jdn5126) [inline cmd/routed-eni-cni-plugin/cni.go:285]: At this point in time, we have already created a veth pair for the pod, configured ip rules, and marked the IP address in the datastore as in-use. Returning an error will cause the container-runtime to retry the CNI ADD, so are we sure that everything that has already been done is idempotent?

Note that above, if we fail to properly setup the virtual interfaces for the pod, we make a `c.DelNetwork` call before returning an error. I think that same recovery logic needs to apply here, plus maybe more.

File: cmd/routed-eni-cni-plugin/cni_test.go [hunk 3]
```diff
@@ -110,6 +110,92 @@ func TestCmdAdd(t *testing.T) {
 	assert.Nil(t, err)
 }
 
+func TestCmdAddwithNPenabled(t *testing.T) {
```
Comment (jdn5126) [inline cmd/routed-eni-cni-plugin/cni_test.go:113]: nit: "TestCmdAddWithNPenabled"

File: cmd/routed-eni-cni-plugin/cni_test.go [hunk 4]
```diff
@@ -110,6 +110,92 @@ func TestCmdAdd(t *testing.T) {
...
+	mockNP.EXPECT().EnforceNpToPod(gomock.Any(), gomock.Any()).Return(enforceNpReply, nil)
+
+	v4Addr := &net.IPNet{
+		IP:   net.ParseIP(addNetworkReply.IPv4Addr),
+		Mask: net.IPv4Mask(255, 255, 255, 255),
+	}
+	mocksNetwork.EXPECT().SetupPodNetwork(gomock.Any(), cmdArgs.IfName, cmdArgs.Netns,
+		v4Addr, nil, int(addNetworkReply.DeviceNumber), gomock.Any(), gomock.Any()).Return(nil)
+
+	err := add(cmdArgs, mocksTypes, mocksGRPC, mocksRPC, mocksNetwork)
```
Comment (jdn5126) [inline cmd/routed-eni-cni-plugin/cni_test.go:195]: This unit test needs to cover cleanup of any resources allocated as a result of the successful `AddNetworkReply`

File: pkg/ipamd/ipamd.go [hunk 5]
```diff
@@ -1731,6 +1737,13 @@ func enablePodENI() bool {
 	return utils.GetBoolAsStringEnvVar(envEnablePodENI, false)
 }
 
+func getNetworkPolicyMode() string {
+	if value := os.Getenv(envNetworkPolicyMode); value != "" {
```
Comment (jdn5126) [inline pkg/ipamd/ipamd.go:1741]: nit: This should either log an error on a typo like "strct", or we should convert to a boolean or enum

File: cmd/routed-eni-cni-plugin/cni.go (hunk 1 above)
Comment (jayanthvn) [reply cmd/routed-eni-cni-plugin/cni.go:280]: Yeah this was mainly for testing purpose, have few pending tasks for the PR..

File: cmd/routed-eni-cni-plugin/cni.go (hunk 2 above)
Comment (jayanthvn) [reply cmd/routed-eni-cni-plugin/cni.go:285]: Yes, have a cleanup function to call..

File: pkg/ipamd/ipamd.go (hunk 5 above)
Comment (jayanthvn) [reply pkg/ipamd/ipamd.go:1741]: yes will be adding few logs around this and we would need aws-node to fail..

File: cmd/routed-eni-cni-plugin/cni_test.go (hunk 4 above)
Comment (jayanthvn) [reply cmd/routed-eni-cni-plugin/cni_test.go:195]: Will be handling this once I add the cleanup code..

File: cmd/routed-eni-cni-plugin/cni.go [hunk 6]
```diff
@@ -275,6 +278,34 @@ func add(args *skel.CmdArgs, cniTypes typeswrapper.CNITYPES, grpcClient grpcwrap
 	// dummy interface is appended to PrevResult for use during cleanup
 	result.Interfaces = append(result.Interfaces, dummyInterface)
 
+	if utils.IsStrictMode(r.NetworkPolicyMode) {
+		// Set up a connection to the ipamD server.
```
Comment (achevuru) [inline cmd/routed-eni-cni-plugin/cni.go:282]: `Set up a connection to the ipamD server` --> `Set up a connection to the network policy agent`

Comment (jayanthvn) [inline cmd/routed-eni-cni-plugin/cni.go:282]: nit ipamd->np

File: cmd/routed-eni-cni-plugin/cni.go [hunk 7]
```diff
@@ -275,6 +278,34 @@ func add(args *skel.CmdArgs, cniTypes typeswrapper.CNITYPES, grpcClient grpcwrap
...
+
+		npr, err := npc.EnforceNpToPod(context.Background(),
+			&pb.EnforceNpRequest{
+				K8S_POD_NAME:      string(k8sArgs.K8S_POD_NAME),
+				K8S_POD_NAMESPACE: string(k8sArgs.K8S_POD_NAMESPACE),
+			})
+
+		// No need to cleanup IP and network, kubelet will send delete.
+		if err != nil || !npr.Success {
+			log.Errorf("Failed to setup default network policy Pod Name %s and NameSpace %s: GRPC returned - %v Network policy agent returned - %v",
```
Comment (jayanthvn) [inline cmd/routed-eni-cni-plugin/cni.go:301]: nit -> network policy for...

Comment (achevuru) [inline cmd/routed-eni-cni-plugin/cni.go:301]: nit: network policy **for** Pod Name

File: pkg/ipamd/ipamd.go [hunk 8]
```diff
@@ -176,6 +176,10 @@ const (
 	// aws error codes for insufficient IP address scenario
 	INSUFFICIENT_CIDR_BLOCKS    = "InsufficientCidrBlocks"
 	INSUFFICIENT_FREE_IP_SUBNET = "InsufficientFreeAddressesInSubnet"
+
+	// envEnableNetworkPolicy is used to enable IPAMD/CNI to send pod create updates to network policy agent.
```
Comment (achevuru) [inline pkg/ipamd/ipamd.go:180]: nit: `create updates` --> `create events`

File: pkg/ipamd/ipamd.go [hunk 9]
```diff
@@ -176,6 +176,10 @@ const (
...
+	envNetworkPolicyMode     = "NETWORK_POLICY_ENFORCING_MODE"
+	defaultNetworkPolicyMode = "none"
```
Comment (achevuru) [inline pkg/ipamd/ipamd.go:182]: `none` --> `standard`

File: cmd/routed-eni-cni-plugin/cni.go [hunk 10]
```diff
@@ -275,6 +278,34 @@ func add(args *skel.CmdArgs, cniTypes typeswrapper.CNITYPES, grpcClient grpcwrap
...
+		//Make a GRPC call for network policy agent
+		npc := rpcClient.NewNPBackendClient(npConn)
+
+		npr, err := npc.EnforceNpToPod(context.Background(),
+			&pb.EnforceNpRequest{
+				K8S_POD_NAME:      string(k8sArgs.K8S_POD_NAME),
+				K8S_POD_NAMESPACE: string(k8sArgs.K8S_POD_NAMESPACE),
+			})
+
+		// No need to cleanup IP and network, kubelet will send delete.
```
Comment (jdn5126) [inline cmd/routed-eni-cni-plugin/cni.go:299]: Why would kubelet send a delete? Won't it just retry the ADD?

Comment (jayanthvn) [reply cmd/routed-eni-cni-plugin/cni.go:299]: We can discuss offline more, if Add fails then kubelet will send a delete.... but check this out -

A runtime should not call ADD twice (without an intervening DEL) for the same (CNI_CONTAINERID, CNI_IFNAME) tuple. This implies that a given container ID may be added to a specific network more than once only if each addition is done with a different interface name.

https://www.cni.dev/docs/spec/#add-add-container-to-network-or-apply-modifications

I have attached an example as well in the test..

**Review by achevuru — APPROVED** [review]

**Pull request merged**